update_existing_families.py

```diff
@@ -0,0 +1,146 @@
+    expected_set = set(expected)
+    actual_set = set(actual)
+    missing_list = list(expected_set.difference(actual_set))
+    if len(missing_list) > 0:
+        print("Missing expected header(s) in %s: %s" % (name, missing_list))
+        sys.exit(1)
+    extra_list = list(actual_set.difference(expected_set))
+    if len(extra_list) > 0:
+        print("Found unexpected header(s) in %s: %s" % (name, extra_list))
+        sys.exit(1)
+
+# Identify existing families:
+# - Update student information particularly if they have changed school.
+# - Add any new students to the existing families
+# - Update any students who are no longer in BSD (set school="No Longer in BSD").
+# - If possible, update the address of existing donor/families if they have moved.
+# - If possible, update any email changes. 
+
+# Mapping of district school name to dp school code
+DISTRICT_SCHOOL_MAPPING = {
+    'Burlingame Intermediate School': 'BIS',
+    'Franklin Elementary School': 'FRANKLIN',
+    'Hoover Elementary School': 'HOOVER',
+    'Lincoln Elementary School': 'LINCOLN',
+    'McKinley Elementary School': 'MCKINLEY',
+    'Roosevelt Elementary School': 'ROOSEVELT',
+    'Washington Elementary School': 'WASHINGTON',
+}
+
+def district_school_to_dp_school(name):
+    return DISTRICT_SCHOOL_MAPPING[name]
+
+# Load dp data keyed off student number
+dp_records_multidict = defaultdict(list)
+with open(sys.argv[1], 'rb') as dp_inputfile:
+    reader = csv.DictReader(dp_inputfile)
+    validate_headers(DP_REPORT_221_HEADERS, reader.fieldnames, "DonorPerfect export")
+    for row in reader:
+        if len(row['STU_NUMBER']) > 0:
+            dp_records_multidict[row['STU_NUMBER']].append(row)
+        
+# Load district data keyed off student number ("system id" there)
+district_records_dict = {}
+with open(sys.argv[2], 'rb') as district_inputfile:
+    reader = csv.DictReader(district_inputfile)
+    validate_headers(DISTRICT_DATA_HEADERS, reader.fieldnames, "district data")
+    for row in reader:
+        district_records_dict[row['SystemID']] = row
+
+# Make updates for existing students
+dp_import_studentrecords = []
+for student_id in dp_records_multidict:
+    for dp_record in dp_records_multidict[student_id]:
+        if student_id in district_records_dict:
+            # Returning student
+            studentrecord = dp_record.copy()
+            # Update current grade
+            studentrecord['GRADE'] = district_records_dict[student_id]['Grade']
+            # Update school
+            studentrecord['SCHOOL'] = district_school_to_dp_school(district_records_dict[student_id]['School'])
+            dp_import_studentrecords.append(studentrecord)
+        elif dp_record['GRADE'] == '8' and dp_record['SCHOOL'] == 'BIS':
+            # Graduation time
+            studentrecord = dp_record.copy()
+            studentrecord['GRADE'] = '9'
+            studentrecord['SCHOOL'] = 'ALUM'
+            dp_import_studentrecords.append(studentrecord)
+
```

Comment:
Does it make sense to add another branch here for the case where the student is not in the district records and not in 8th grade? Such students have left the district before graduating...